### PR TITLE
CXP-1299: Disable a catalog when removing an attribute option that is in the product selection

### PR DESCRIPTION
**Pull request by guillaumefer-akeneo**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

- [ ] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: components/catalogs/back/src/Application/Persistence/DisableCatalogsQueryInterface.php [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+ * @author    Willy Mesnage <willy.mesnage@akeneo.com>
+ * @copyright 2022 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+interface DisableCatalogsQueryInterface
+{
+    /**
+     * @param array<string> $catalogsUUID
+     */
+    public function execute(array $catalogsUUID): void;
```
Comment (masacc) [inline components/catalogs/back/src/Application/Persistence/DisableCatalogsQueryInterface.php:17]:
```suggestion
    public function execute(array $catalogsUuids): void;
```

File: components/catalogs/back/src/Infrastructure/EventSubscriber/AttributeOptionRemovalSubscriber.php [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+use Akeneo\Pim\Structure\Component\Model\AttributeOptionInterface;
+use Akeneo\Tool\Bundle\BatchBundle\Launcher\JobLauncherInterface;
+use Akeneo\Tool\Component\Batch\Model\JobInstance;
+use Akeneo\Tool\Component\StorageUtils\Repository\IdentifiableObjectRepositoryInterface;
+use Akeneo\Tool\Component\StorageUtils\StorageEvents;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\EventDispatcher\GenericEvent;
+use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
+
+class AttributeOptionRemovalSubscriber implements EventSubscriberInterface
```
Comment (masacc) [inline components/catalogs/back/src/Infrastructure/EventSubscriber/AttributeOptionRemovalSubscriber.php:16]: missing 
```
/**
 * @author    Willy Mesnage <willy.mesnage@akeneo.com>
 * @copyright 2022 Akeneo SAS (http://www.akeneo.com)
 * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
 */
 ```

and in some other files

File: components/catalogs/back/src/Infrastructure/EventSubscriber/AttributeOptionRemovalSubscriber.php [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+    public function disableCatalogsIfAttributeOptionIsRemoved(GenericEvent $event): void
+    {
+        $attributeOption = $event->getSubject();
+        if (!$attributeOption instanceof AttributeOptionInterface) {
+            return;
+        }
+
+        $attributeCode = $attributeOption->getAttribute()->getCode();
+        $attributeOptionCode = $attributeOption->getCode();
```
Comment (masacc) [inline components/catalogs/back/src/Infrastructure/EventSubscriber/AttributeOptionRemovalSubscriber.php:40]: useless variables

File: components/catalogs/back/src/Infrastructure/Job/DisableCatalogsOnAttributeOptionRemovalTasklet.php [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+            throw new \LogicException('the variable $stepExecution should not be null.');
+        }
+
+        /** @var string $attributeCode */
+        $attributeCode = $this->stepExecution->getJobParameters()->get('attribute_code');
+
+        /** @var string $attributeOptionCode */
+        $attributeOptionCode = $this->stepExecution->getJobParameters()->get('attribute_option_code');
+
+        $catalogsUUID = $this->getCatalogsToDisableQuery->execute($attributeCode, $attributeOptionCode);
```
Comment (masacc) [inline components/catalogs/back/src/Infrastructure/Job/DisableCatalogsOnAttributeOptionRemovalTasklet.php:39]:
```suggestion
        $catalogsUuids = $this->getCatalogsToDisableQuery->execute($attributeCode, $attributeOptionCode);
```

**Review by masacc — APPROVED** [review]

**Comment by tseho** [comment]
You need to update the fixtures in EE too.

File: components/catalogs/back/src/Application/Persistence/DisableCatalogsQueryInterface.php (hunk 1 above)
Comment (RiffVentura) [inline components/catalogs/back/src/Application/Persistence/DisableCatalogsQueryInterface.php:17]:
```suggestion
    public function execute(array $id): void;
```
As it is done for other queries related to catalog directly

File: components/catalogs/back/src/Infrastructure/Persistence/GetCatalogsToDisableOnAttributeOptionRemovalQuery.php [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+namespace Akeneo\Catalogs\Infrastructure\Persistence;
+
+use Akeneo\Catalogs\Application\Persistence\GetCatalogsToDisableOnAttributeOptionRemovalQueryInterface;
+use Doctrine\DBAL\Connection;
+
+/**
+ * @copyright 2022 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+final class GetCatalogsToDisableOnAttributeOptionRemovalQuery implements GetCatalogsToDisableOnAttributeOptionRemovalQueryInterface
```
Comment (RiffVentura) [inline components/catalogs/back/src/Infrastructure/Persistence/GetCatalogsToDisableOnAttributeOptionRemovalQuery.php:14]: Technically speaking it only finds catalog ids that contain an attribute option
I think the naming should reflect that :
`GetCatalogIdsContainingAttributeOption`

**Review by RiffVentura — APPROVED** [review]
I might be wrong but I think it needs a migration for the new job instance

**Comment by guillaumefer-akeneo** [comment]
This PR is abandoned, the new simpler version will be in this branch : CXP-1306

**Pull request closed without merging**